flow/BlobCipher.cpp

```diff
@@ -129,6 +129,9 @@ BlobCipherKeyIdCacheKey BlobCipherKeyIdCache::getCacheKey(const EncryptCipherBas
 }
 
 Reference<BlobCipherKey> BlobCipherKeyIdCache::getLatestCipherKey() {
+	if (latestBaseCipherKeyId == ENCRYPT_INVALID_CIPHER_KEY_ID) {
+		return Reference<BlobCipherKey>();
```

Comment:
we should be throwing an exception here; this should be a code logic bug?

Reply:
ENCRYPT_INVALID_CIPHER_KEY_ID exactly means the cache is empty.

Reply:
I think it is better not to conflate "invalid_cipher_id" with the "cache empty", a caller can pass "invalid_id" due to some bug even when cache is not-empty. We can add a helper method in the cache to return its size if needed.

Reply:
with getLatestCipherKey the caller is not passing in key id, so the ENCRYPT_INVALID_CIPHER_KEY_ID indicate internal state (no latest key). I agree for getCipherKey, it should come with invalid id check, which I'm not updating in this PR.

Reply:
If the cache is empty, then BlobCipherKeyIdCache should be invoked? 

```
	auto domainItr = domainCacheMap.find(domainId);
	if (domainItr == domainCacheMap.end()) {
		TraceEvent("GetLatestCipherKey_DomainNotFound").detail("DomainId", domainId);
		throw encrypt_key_not_found();
	}
```
The domainId check should return empty_reference based on your change? Am I missing anything here.

Reply:
Looking again, for empty cache case, even with suggested modification, the code should throw "key_not_found" from BlobCipherKeyCache::getLatestCipher(domainId) method. 

However, if the "domainId" entry exists in the KeyCache, then 'latestCipherBaseId" can't be INVALID, I think

Reply:
Right. This is actually for the valgrind fix. Somewhere in EncryptionOps test it access the BlobCipherKeyIdCache directly, and when the cache is empty, the TraceEvent code will try to print the uninitialized salt. Instead of zero-out the salt on initialization, I choose to make the code not touching the salt in this case.

Reply:
Got it!
I don't this it is a good idea to put this code here to solve the issue at hand, it would be better to find the test code and fix it. Let me know if you want me to look more into the Valgrind issue (if that is impacting your open items task progress). 
Thanks

Reply:
For the empty cache case, I prefer to not throwing "key_not_found" because this is not actually an error, and a try-catch on caller side looks cumbersome. What do you think?

Reply:
Ah, sorry. I look at the valgrind stack again, and it is not EncryptionOps accessing BlobCipherKeyIdCache directly. It is BlobCipherKeyIdCache::insertBaseCipherKey calling getLatestCipherKey() before actually insert the key. In this case the BlobCipherKeyIdCache is created, but it is empty. In this case, does it make sense to keep the existing change?

Valgrind stack:
```
==46== Conditional jump or move depends on uninitialised value(s)
==46==    at 0x6E09DAE: vfprintf (in /usr/lib64/libc-2.17.so)
==46==    by 0x6E37178: vsnprintf (in /usr/lib64/libc-2.17.so)
==46==    by 0x37DDB51: vsformat(std::string&, char const*, __va_list_tag*) (flow.cpp:229)
==46==    by 0x37DE06C: format(char const*, ...) (flow.cpp:266)
==46==    by 0xC00D07: toString (Trace.h:255)
==46==    by 0xC00D07: std::enable_if<Traceable<unsigned long>::value, BaseTraceEvent&>::type BaseTraceEvent::detail<unsigned long>(char const*, unsigned long const&) (Trace.h:434)
==46==    by 0x36D5BF6: BlobCipherKeyIdCache::getCipherByBaseCipherId(unsigned long const&, unsigned long const&) (BlobCipher.cpp:147)
==46==    by 0x36D95A7: getLatestCipherKey (BlobCipher.cpp:137)
==46==    by 0x36D95A7: BlobCipherKeyIdCache::insertBaseCipherKey(unsigned long const&, unsigned char const*, int) (BlobCipher.cpp:161)
==46==    by 0x36D9E0B: BlobCipherKeyCache::insertCipherKey(long const&, unsigned long const&, unsigned char const*, int) (BlobCipher.cpp:253)
==46==    by 0x23EC77E: EncryptionOpsWorkload::setupCipherEssentials() (EncryptionOps.actor.cpp:178)
==46==    by 0x23ED0A5: EncryptionOpsWorkload::start(Database const&) (EncryptionOps.actor.cpp:305)
==46==    by 0x1E5F2EA: CompoundWorkload::start(Database const&) (tester.actor.cpp:301)
==46==    by 0x1E2E193: (anonymous namespace)::RunWorkloadAsyncActorState<(anonymous namespace)::RunWorkloadAsyncActor>::a_body1loopBody1when2(ReplyPromise<Void>&&, int) (tester.actor.cpp:592)
```

Reply:
> by 0x36D95A7: BlobCipherKeyIdCache::insertBaseCipherKey(unsigned long const&, unsigned char const*, int) (BlobCipher.cpp:161)
```

		if (cipherKey->getBaseCipherId() == baseCipherId) {
			if (memcmp(cipherKey->rawBaseCipher(), baseCipher, baseCipherLen) == 0) {
				TraceEvent("InsertBaseCipherKey_AlreadyPresent")
				    .detail("BaseCipherKeyId", baseCipherId)
				    .detail("DomainId", domainId);                     <---- line 161
				// Key is already present; nothing more to do.
```
Something doesn't add up; the above scenario is ONLY possible when caller attempts to insert a "duplicate" baseCipherId with "exact CipherKey" match usecase (hence, no tampering or foul play). 

I think the problem is the following line of code - 

```
EncryptionOps.actor.cpp::setupCipherEssentials()
...
		// insert the Encrypt Header cipherKey
		generateRandomBaseCipher(AES_256_KEY_LENGTH, &buff[0], &cipherLen);
		cipherKeyCache->insertCipherKey(
		    ENCRYPT_HEADER_DOMAIN_ID, headerBaseCipherId, buff, cipherLen, headerRandomSalt)
...
```

As expected inserting a cipher with API accepting 'salt' shouldn't update 'latestBaseCipher' details (as this entry should happen as part of reading encyrption header), however, the testcode is doing it as a 'hack' for "EncryptHeader authToken cipherKey". 

Then,  below code simulates "cache-miss" randomly for any {"domainId", "baseCipherId"} entry, and when done for domainId == -2 (EncryptHeader auth cipher key), the TraceEvent goes bonkers :)

```
Reference<BlobCipherKey> getEncryptionKey(const EncryptCipherDomainId& domainId,
	                                          const EncryptCipherBaseKeyId& baseCipherId,
	                                          const EncryptCipherRandomSalt& salt) {
		const bool simCacheMiss = deterministicRandom()->randomInt(1, 100) < 15;

		Reference<BlobCipherKeyCache> cipherKeyCache = BlobCipherKeyCache::getInstance();
		Reference<BlobCipherKey> cipherKey = cipherKeyCache->getCipherKey(domainId, baseCipherId, salt);

		if (simCacheMiss) {
			TraceEvent("SimKeyCacheMiss").detail("EncyrptDomainId", domainId).detail("BaseCipherId", baseCipherId);
			// simulate KeyCache miss that may happen during decryption; insert a CipherKey with known 'salt'
			cipherKeyCache->insertCipherKey(domainId,
			                                baseCipherId,
			                                cipherKey->rawBaseCipher(),
			                                cipherKey->getBaseCipherLen(),
			                                cipherKey->getSalt());
			// Ensure the update was a NOP
			Reference<BlobCipherKey> cKey = cipherKeyCache->getCipherKey(domainId, baseCipherId, salt);
			ASSERT(cKey->isEqual(cipherKey));
		}
		return cipherKey;
	}
```

I think I know how to fix this issue. Please let me know if it OK to leave the Valgrind fix from this patch and I will fix it separately? 

Thanks again for debugging this issue :).

Reply:
Sure. Go ahead with the fix. I will close the current PR and subsume it into my upcoming PR.

Reply:
> > A try-catch on caller side looks cumbersome. What do you think?
> 
> My initial thought that it is better for the caller to handle exceptions from a caching library, some exceptions can be 'acceptable' such as: _key_not_found_ which could be satisfied by reaching out to KMS via EKPServer, however, I see below scheme could work as well:
> 
> 1. BlobCipherKeyCache throws _key_not_found_ only for unexpected scenarios such as the one we discussed above. For a _cache_miss_ it is OK to return an emptyReference. However, it is better to have unified behavior of empty-cache as well as cache-missing-a -specific scenario. If we choose not to throw exception, then I would say better not to distinguish these cases as technically they are same in some ways.
> 2. As FDB EKs have 1:1 relation with KMS EKs, we can update EncryptKeyServer lookup interface to ensure that all cipher-keys requested are returned back to the client; if not, server throws "encrypt_key_not_found" (never to happen scenario and points to code logic bug/tampering etc.)
> 
> #2 I can take care as part of #6898.
> 
> Does this sounds reasonable? Thanks

Make sense to me. I will have #1 done in my upcoming PR.